library/common/engine.cc

```diff
@@ -112,18 +121,18 @@ Engine::~Engine() {
   {
     Thread::LockGuard lock(mutex_);
 
-    if (!main_common_) {
+    if (!event_dispatcher_) {
       cv_.wait(mutex_);
     }
 
-    ASSERT(main_common_);
+    ASSERT(event_dispatcher_);
 
     // Exit the event loop and finish up in Engine::run(...)
     event_dispatcher_->exit();
```

Comment:
the joinable call above should be updated, as it doesn't mean what we thought. And do we want to ultimately allow for exit from the main_thread's context? Per my experiments exiting from that context results in the event loop never actually exiting.